src/libxtp/gwbse/sigma_cda.cc

```diff
@@ -0,0 +1,120 @@
+namespace xtp {
+
+void Sigma_CDA::PrepareScreening() {
+  GaussianQuadrature::options opt;
+  opt.homo = _opt.homo;
+  opt.order = _opt.order;
+  opt.qptotal = _qptotal;
+  opt.qpmin = _opt.qpmin;
+  opt.rpamin = _opt.rpamin;
+  opt.alpha = _opt.alpha;
+  opt.quadrature_scheme = _opt.quadrature_scheme;
+  _gq.configure(opt, _rpa);
+}
+
+// This function is used in the calculation of the residues
+// This calculates eps^-1 (inverse of the dielectric function) for complex
+// frequencies of the kind omega = delta + i*eta
+double Sigma_CDA::CalcDiagContribution(Eigen::RowVectorXd Imx_row, double delta,
+                                      double eta) const {
+  Eigen::MatrixXcd DielMxInv =
+      _rpa.calculate_epsilon_complex(delta, eta).inverse();
+  DielMxInv.diagonal().array() -= 1.0;
+  std::complex<double> lambda =
+      ((Imx_row * DielMxInv).cwiseProduct(Imx_row)).sum();
+  return lambda.real();
+}
+
+double Sigma_CDA::CalcResiduePrefactor(double e_f, double e_m,
+                                      double frequency) const {
+  double factor = 0.0;
+
+  if (e_f < e_m && e_m < frequency) {
+    factor = 1.0;
+  } else if (e_f > e_m && e_m > frequency) {
+    factor = -1.0;
+  } else if (e_m - frequency == 0.0 && e_f > e_m) {
+    factor = -0.5;
+  } else if (e_m - frequency == 0.0 && e_f < e_m) {
+    factor = 0.5;
+  }
+  return factor;
+}
+
+double Sigma_CDA::CalcResidueContribution(Eigen::VectorXd rpa_energies,
+                                         double frequency,
+                                         Index gw_level) const {
+  Index rpatotal = rpa_energies.size();
+  double sigma_c = 0.0;
+  double result_alpha = 0.0;
+  Index homo = _opt.homo - _opt.rpamin;
+  Index lumo = homo + 1;
+  double fermi_rpa = (rpa_energies(lumo) + rpa_energies(homo)) / 2.0;
+
+  const Eigen::MatrixXd& Imx = _Mmn[gw_level];
+  for (Index i = 0; i < rpatotal; ++i) {
+    double delta = std::abs(rpa_energies(i) - frequency);
+    double factor = CalcResiduePrefactor(fermi_rpa, rpa_energies(i), frequency);
+
+    // just sum contributing terms
+    if (std::abs(factor) > 0.1) {
+      sigma_c += factor * CalcDiagContribution(Imx.row(i), delta, _eta);
+    }
+  }
+  return sigma_c;
+}
+
+double Sigma_CDA::CalcCorrelationDiagElement(Index gw_level,
+                                            double frequency) const {
+
+  const Eigen::VectorXd& RPAenergies = _rpa.getRPAInputEnergies();
+
+  double sigma_c_residue =
+      CalcResidueContribution(RPAenergies, frequency, gw_level);
+
+  double sigma_c_integral = _gq.SigmaGQDiag(frequency, gw_level, _opt.alpha);
+
+  return sigma_c_residue + sigma_c_integral;
+}
+
+double Sigma_CDA::CalcDiagContributionValue_alpha(Eigen::RowVectorXd Imx_row,
+                                                 double delta,
+                                                 double alpha) const {
+  Eigen::MatrixXcd R = _rpa.calculate_epsilon_complex(0.0, 0.0).inverse();
+
+  R.diagonal().array() -= 1.0;
```

Comment:
I do not know how often you have to calculate it, but this part is independent of the input arguments of this function. Do you maybe have to calculate this only once?

Reply:
Not needed because I removed the alpha dependencies